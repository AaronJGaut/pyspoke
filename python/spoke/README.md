# `spoke`

A python library supporting pubsub and remote procedure calls using JSON messages.

# Isn't this just a bad dbus?

Actually this project started out as a bad ROS, but yes.

# Installation

Why?
