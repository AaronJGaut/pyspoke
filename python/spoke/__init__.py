from . import genid
from . import serialize
from . import client
from . import routing
from . import server
from . import sync
from . import connection
