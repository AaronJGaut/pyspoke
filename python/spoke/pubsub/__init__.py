from . import client
from . import msgpack
from . import route
from . import server
