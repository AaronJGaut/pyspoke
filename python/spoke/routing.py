def tokenize(channel):
    tokens = [t for t in channel.split("/") if t]
    change = True
    # In runs of * and ** tokens, all the * tokens should come first to simplify matching
    is_wc = [t in ("*", "**") for t in tokens]
    edges = [i + 1 for i, (l, r) in enumerate(zip(is_wc[:-1], is_wc[1:])) if l != r]
    run_idxs = [0] + edges + [len(tokens)]
    runs = [tokens[start:end] for start, end in zip(run_idxs[:-1], run_idxs[1:])]
    for run in runs:
        if "**" not in run:
            continue
        wc1_count = sum([x == "*" for x in run])
        run.clear()
        run.extend(["*"] * wc1_count)
        run.append("**")
    return tuple(sum(runs, []))


class Entry:
    def __init__(self, tokens, sub, sub_id=None):
        self.tokens = tokens
        self.sub = sub
        self.sub_id = sub_id or sub

    def __eq__(self, other):
        return self.tokens == other.tokens and self.sub_id == other.sub_id

    def __hash__(self):
        return hash((hash(self.tokens) * 73) ^ (hash(self.sub_id) * 71))

    def test(self, tokens):
        stack = [(0, 0)]
        while stack:
            ci, ri = stack.pop()
            if ci >= len(tokens) and ri >= len(self.tokens):
                return True
            if ri >= len(self.tokens):
                continue
            rt = self.tokens[ri]
            if ri == len(self.tokens) - 1 and rt == "**":
                return True
            if ci >= len(tokens):
                continue
            ct = tokens[ci]
            if rt == "*" or rt == ct:
                stack.append((ci + 1, ri + 1))
                continue
            if rt == "**":
                lrt = self.tokens[ri + 1]
                for lci, lct in enumerate(tokens[ci:]):
                    if lrt == lct:
                        stack.append((lci + ci, ri + 1))
                continue
        return False

class SubscriberTable:
    def __init__(self):
        self.entries = set()

    def add_rule(self, channel, sub, sub_id=None):
        tokens = tokenize(channel)
        entry = Entry(tokens, sub, sub_id)
        self.entries.add(entry)

    def remove_rule(self, channel, sub_id):
        tokens = tokenize(channel)
        entry = Entry(tokens, None, sub_id)
        if entry in self.entries:
            self.entries.remove(entry)

    def remove_sub(self, sub_id):
        condemned = []
        for entry in self.entries:
            if entry.sub_id == sub_id:
                condemned.append(entry)
        for x in condemned:
            self.entries.remove(x)

    def get_subs(self, channel):
        tokens = tokenize(channel)
        subs = set()
        for entry in self.entries:
            if entry.sub in subs:
                continue
            if entry.test(tokens):
                subs.add(entry.sub)
        return list(subs)
