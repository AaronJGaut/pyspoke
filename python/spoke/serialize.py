import json
def msg_to_bytes(channel, msg=None):
    full_msg = {
        "channel": channel,
        "payload": msg,
    }
    msg_str = json.dumps(full_msg)
    return (msg_str).encode("utf8") + b"\0"

def bytes_to_msg(msg_data):
    full_msg = json.loads(msg_data[:-1].decode("utf8"))
    channel = full_msg.get("channel", None)
    msg = full_msg.get("payload", None)
    return channel, msg

