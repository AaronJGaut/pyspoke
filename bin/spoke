#!/usr/bin/env python
import asyncio
from spoke.pubsub.server import Server

server = Server()

try:
    asyncio.run(server.run())
except KeyboardInterrupt:
    pass
