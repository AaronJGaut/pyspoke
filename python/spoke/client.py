import os
import asyncio
from spoke import serial

class Client:
    def __init__(self, host=None, port=None):
        if host is None:
            host = os.getenv("SPOKEHOST", None) or "localhost"
        if port is None:
            port = os.getenv("SPOKEPORT", None) or 8888

        self.connected = False
        self.reader = None
        self.writer = None
        self.port = port
        self.host = host
        self.subs = {}

    async def publish(self, channel, msg):
        msg_data = serial.msg_to_bytes(channel, msg)
        self.writer.write(msg_data)
        await self.writer.drain()

    async def subscribe(self, channel, callback):
        if channel not in self.subs:
            await self.publish("spoke_control", channel)
        self.subs[channel] = callback

    async def unsubscribe(self, channel):
        # TODO: tell server
        if channel in self.subs:
            del self.subs[channel]

    async def run(self, wait=False):
        reader, writer = await asyncio.open_connection(self.host, self.port)
        self.connected = True
        self.reader = reader
        self.writer = writer
        asyncio.create_task(self._listen())
        if wait:
            await self.wait()

    async def wait(self):
        while self.connected:
            await asyncio.sleep(1)

    async def _listen(self):
        while True:
            try:
                msg_data = await self.reader.readuntil(separator=b"\0")
            except asyncio.exceptions.IncompleteReadError:
                self.connected = False
                break
            channel, msg = serial.bytes_to_msg(msg_data)
            if channel in self.subs:
                await self.subs[channel](msg)
