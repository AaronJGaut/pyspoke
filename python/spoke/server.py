import os
import asyncio
from spoke import serial

class ClientHandle:
    def __init__(self, reader, writer, server):
        self.reader = reader
        self.writer = writer
        self.server = server

    async def handle_recv(self):
        while True:
            try:
                msg_data = await self.reader.readuntil(separator=b'\0')
            except asyncio.exceptions.IncompleteReadError:
                # TODO: unsubscribe all
                break
            channel, msg = serial.bytes_to_msg(msg_data)
            if channel == "spoke_control" and msg:
                self.server.subscribe(self, msg)
            if channel:
                await self.server.publish(channel, msg)

    async def send(self, channel, msg):
        msg_data = serial.msg_to_bytes(channel, msg)
        self.writer.write(msg_data)
        await self.writer.drain()

class Server:
    def __init__(self, port=None):
        if port is None:
            port = os.getenv("SPOKEPORT", None) or 8888
        self._server = None
        self._subs = {}
        self._port = port

    def get_subs(self, channel):
        return list(self._subs.get(channel, []))

    def subscribe(self, client, channel):
        if channel not in self._subs:
            self._subs[channel] = {client}
        else:
            self._subs[channel].add(client)

    def unsubscribe(self, client, channel):
        if channel in self._subs:
            self._subs[channel] -= {client}

    async def publish(self, channel, msg):
        subs = self.get_subs(channel)
        for sub in subs:
            await sub.send(channel, msg)

    async def run(self):
        self._server = await asyncio.start_server(self.handle_conn, "0.0.0.0", self._port)
        async with self._server:
            await self._server.serve_forever()

    async def handle_conn(self, reader, writer):
        client = ClientHandle(reader, writer, self)
        asyncio.create_task(client.handle_recv())
