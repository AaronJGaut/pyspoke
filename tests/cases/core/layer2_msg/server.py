import os
import asyncio
import spoke


clients = []

async def handle_client(conn):
    clients.append(conn)
    print("server.handle_connect")
    try:
        async for msg in conn:
            print(f"server.handle_recv: {msg}")
            msg["value"] += 10
            for client in clients:
                await client.send(msg)
    except ConnectionError:
        pass
    finally:
        print("server.handle_disconnect")
        clients.remove(conn)

async def main():
    async with spoke.conn.pack.Server(
        conn_server_class = spoke.conn.socket.Server,
        packer_class = spoke.conn.pack.JsonPacker,
        conn_opts = {"reuse": True},
    ) as server:
        async for client in server:
            asyncio.create_task(handle_client(client))

try:
    asyncio.run(main())
except KeyboardInterrupt:
    pass
