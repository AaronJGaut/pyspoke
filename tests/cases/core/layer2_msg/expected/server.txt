server.handle_connect
server.handle_recv: {'name': 'client1', 'value': 0}
server.handle_recv: {'name': 'client1', 'value': 1}
server.handle_recv: {'name': 'client1', 'value': 2}
server.handle_connect
server.handle_recv: {'name': 'client2', 'value': 0}
server.handle_recv: {'name': 'client1', 'value': 3}
server.handle_recv: {'name': 'client2', 'value': 1}
server.handle_recv: {'name': 'client1', 'value': 4}
server.handle_recv: {'name': 'client2', 'value': 2}
server.handle_disconnect
server.handle_recv: {'name': 'client2', 'value': 3}
server.handle_recv: {'name': 'client2', 'value': 4}
server.handle_disconnect
